text_generation/causal_lm/cpp/beam_search_causal_lm.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <group_beam_searcher.hpp>
+#include <openvino/openvino.hpp>
+#include <openvino_extensions/strings.hpp>
+
+namespace {
+std::pair<ov::Tensor, ov::Tensor> tokenize(ov::InferRequest&& tokenizer, std::string_view prompt) {
+    ov::Tensor destination = tokenizer.get_input_tensor();
+    openvino_extensions::pack_strings(std::array{prompt}, destination);
+    tokenizer.infer();
+    return {tokenizer.get_tensor("input_ids"), tokenizer.get_tensor("attention_mask")};
+}
+
+std::string detokenize(ov::InferRequest& detokenizer, const std::vector<int64_t>& tokens) {
+    constexpr size_t BATCH_SIZE = 1;
+    ov::Tensor inp = detokenizer.get_input_tensor();
+    inp.set_shape({BATCH_SIZE, tokens.size()});
+    for (size_t idx = 0; idx < tokens.size(); ++idx) {
+        inp.data<int64_t>()[idx] = tokens.at(idx);
+    }
+    detokenizer.infer();
+    return openvino_extensions::unpack_strings(detokenizer.get_output_tensor()).front();
+}
+}
+
+int main(int argc, char* argv[]) try {
+    if (argc != 5) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0]
+            + " <openvino_model.xml> <tokenizer.xml> <detokenizer.xml> '<prompt>'");
+    }
+    ov::Core core;
+    core.add_extension(USER_OV_EXTENSIONS_PATH);  // USER_OV_EXTENSIONS_PATH is defined in root CMakeLists.txt
+    auto [input_ids, mask] = tokenize(core.compile_model(argv[2], "CPU").create_infer_request(), argv[4]);
+    ov::InferRequest detokenizer = core.compile_model(argv[3], "CPU").create_infer_request();
+    ov::InferRequest ireq = core.compile_model(argv[1], "CPU").create_infer_request();
+    ireq.set_tensor("input_ids", input_ids);
+    ireq.set_tensor("attention_mask", mask);
+    ov::Tensor position_ids = ireq.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    ireq.get_tensor("beam_idx").set_shape({1});
+    ireq.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+    Parameters parameters;
+    const int64_t* prompt_data = input_ids.data<const int64_t>();
+    parameters.prompt = std::vector<int64_t>{prompt_data, prompt_data + input_ids.get_size()};
+    GroupBeamSearcher group_beam_searcher{parameters};
+    std::vector<int64_t> next_tokens;
+    std::vector<int32_t> next_beams;
```

Comment:
Please break this wall of code to more organized stages with brief comments. When looking at this region, I just don't want to read it.

Reply:
I came up with a better idea instead of comments. Function names are scoped comments, so I use them